examples/dwarf-validate.rs

```diff
@@ -0,0 +1,249 @@
+            }
+        };
+        let mut entries = unit.entries(&abbrevs);
+        let mut unit_refs = Vec::new();
+        loop {
+            let (_, entry) = match entries.next_dfs() {
+                Err(e) => {
+                    w.error(format!("Invalid DIE for unit {:#x}: {}",
+                                    unit.offset().0,
+                                    error::Error::description(&e)));
+                    return ret;
+                },
+                Ok(None) => break,
+                Ok(Some(entry)) => entry,
+            };
+            ret.die_offsets.push(entry.offset());
+
+            let mut attrs = entry.attrs();
+            loop {
+                let attr = match attrs.next() {
+                    Err(e) => {
+                        w.error(format!("Invalid attribute for unit {:#x} at DIE {:#x}: {}",
+                                        unit.offset().0, entry.offset().0,
+                                        error::Error::description(&e)));
+                        return ret;
+                    },
+                    Ok(None) => break,
+                    Ok(Some(attr)) => attr,
+                };
+                match attr.value() {
+                    AttributeValue::UnitRef(offset) => {
+                        unit_refs.push((entry.offset(), offset));
+                    },
+                    AttributeValue::DebugInfoRef(offset) => {
+                        ret.global_die_references.push((entry.offset(), offset));
+                    }
+                    _ => (),
+                }
+            }
+        }
+        ret.internally_valid = true;
+        ret.die_offsets.shrink_to_fit();
+        ret.global_die_references.shrink_to_fit();
+
+        // Check intra-unit references
+        for (from, to) in unit_refs {
+            if let Err(_) = ret.die_offsets.binary_search(&to) {
+                w.error(format!("Invalid intra-unit reference in unit {:#x} from DIE {:#x} to {:#x}",
+                                unit.offset().0, from.0, to.0));
+            }
+        }
+
+        ret
+    };
+    let processed_units = units.into_par_iter().map(process_unit).collect::<Vec<_>>();
+
+    let check_unit = |summary: &UnitSummary| {
+        if !summary.internally_valid {
+            return;
+        }
+        for &(from, to) in summary.global_die_references.iter() {
+            let u = match processed_units.binary_search_by_key(&to, |v| v.offset) {
+                Ok(i) => &processed_units[i],
+                Err(i) => if i > 0 {
+                        &processed_units[i - 1]
+                    } else {
+                        w.error(format!("Invalid cross-unit reference in unit {:#x} from DIE {:#x} to global DIE {:#x}: no unit found",
+                                        summary.offset.0, from.0, to.0));
+                        continue;
+                    },
+            };
+            if !u.internally_valid {
+                return;
```

Comment:
Should this be a `continue`?

Just curious, did you encounter any units that weren't internally valid?

Reply:
Yes, that should be `continue`. I'll fix it. Thanks!

No, we haven't encountered such files in the wild yet.